### PR TITLE
NetSim base64

### PR DESCRIPTION
**Pull request by Hamms**
adds a serialization layer to NetSimMessages; row payloads that were previously transmitted to DynamoDB as binary strings (which in JavaScript are literally just Strings that happen to contain zeros and ones) are now base64-encoded and much, much smaller.

### CODE REVIEW

File: apps/src/netsim/dataConverters.js [hunk 1]
```diff
@@ -16,10 +16,20 @@
...
+/* global window */
+/* global require */
+/* global exports */
 
 require('../utils'); // For String.prototype.repeat polyfill
 var netsimUtils = require('./netsimUtils');
 
+// window.{btoa, atob} polyfills
+if (!(window.atob && window.btoa)) {
+    var base64 = require('./base64');
```
Comment (bcjordan) [inline apps/src/netsim/dataConverters.js]: nit: double indentation

File: apps/src/netsim/dataConverters.js [hunk 2]
```diff
@@ -16,10 +16,20 @@
...
+    window.btoa = window.btoa || base64.encode;
+    window.atob = window.btoa || base64.decode;
```
Comment (bcjordan) [inline apps/src/netsim/dataConverters.js]: Should this `btoa` be `atob`?

Comment (Hamms) [inline apps/src/netsim/dataConverters.js]: very yes

File: apps/src/netsim/NetSimMessage.js [hunk 3]
```diff
@@ -59,7 +62,8 @@ var NetSimMessage = module.exports = function (shard, messageRow) {
    * All other message content, including the 'packets' students will send.
    * @type {*}
    */
-  this.payload = messageRow.payload;
+  var b64 = messageRow.base64Payload;
+  this.payload = (b64) ? base64ToBinary(b64.string, b64.len) : undefined;
```
Comment (islemaster) [inline apps/src/netsim/NetSimMessage.js]: Style nits:
* Characters are cheap and our stuff gets minified anyway; maybe `s/b64/base64Payload`
* Parentheses around b64 in your ternary expression are unnecessary.

Comment (Hamms) [inline apps/src/netsim/NetSimMessage.js]: I'd argue that the parens, although definitely unnecessary, make it a bit clearer that a ternary is being used

File: apps/src/netsim/NetSimMessage.js [hunk 4]
```diff
@@ -128,7 +132,7 @@ NetSimMessage.prototype.buildRow = function () {
     fromNodeID: this.fromNodeID,
     toNodeID: this.toNodeID,
     simulatedBy: this.simulatedBy,
-    payload: this.payload,
+    base64Payload: this.payload ? binaryToBase64(this.payload) : undefined,
```
Comment (islemaster) [inline apps/src/netsim/NetSimMessage.js]: Please change the messageRow `@typedef` about ten lines above this as well.

File: apps/src/netsim/NetSimRouterNode.js [hunk 5]
```diff
@@ -520,16 +512,16 @@ NetSimRouterNode.prototype.recalculateSchedule = function () {
 
   var queueSizeInBits = 0;
   var pessimisticCompletionTime = this.simulationTime_;
-  var queuedRow;
+  var queuedMsg;
```
Comment (islemaster) [inline apps/src/netsim/NetSimRouterNode.js]: Nit-picking again :grin: `s/queuedMsg/queuedMessage`.  NetSim code is verbose.

File: apps/src/netsim/NetSimRouterNode.js [hunk 6]
```diff
@@ -1418,7 +1407,7 @@ NetSimRouterNode.prototype.getCurrentDataRate = function () {
 /**
  * When the message table changes, we might have a new message to handle.
  * Check for and handle unhandled messages.
- * @param {messageRow[]} rows
+ * @param {message[]} rows
```
Comment (islemaster) [inline apps/src/netsim/NetSimRouterNode.js]: This doc change looks like a mistake; I think `onMessageTableChange_` still takes `messageRow[]`.

Comment (Hamms) [inline apps/src/netsim/NetSimRouterNode.js]: right you are, thanks!

File: apps/src/netsim/base64.js [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/* base64 encode/decode compatible with window.btoa/atob
```
Comment (islemaster) [inline apps/src/netsim/base64.js]: Although this file is only used by NetSim, we should probably put it somewhere else.  I'm not 100% sure what the right place is though.
* At least moving it up a directory wouldn't be bad - we've done that a few times.
* Making an `apps/src/vendor` directory for external code we pull in (we have something similar in dashboard)
* In some cases we prefer to depend on a node module, like [one of these](https://www.npmjs.com/search?q=base64).  For an example of this, see how [momentjs](http://momentjs.com/) is used in NetSimLogEntry and provided via apps/package.json.

File: apps/test/netsim/NetSimMessage.js [hunk 8]
```diff
@@ -94,17 +131,32 @@ describe("NetSimMessage", function () {
         assert(err === null, "Error is null on success");
       });
     });
+
+    it ("throws an exception when given a non-binary String as a payload", function () {
```
Comment (islemaster) [inline apps/test/netsim/NetSimMessage.js]: :+1: :+1: :+1:

File: apps/test/netsim/NetSimRouterNode.js [hunk 9]
```diff
@@ -1449,8 +1449,8 @@ describe("NetSimRouterNode", function () {
           return p + c;
         }, 0);
         assert(droppedPackets === expectedDropCount, "Expected that " +
-        expectedDropCount + " packets would be dropped, " +
-        "but logs only report " + droppedPackets + " dropped packets");
+          expectedDropCount + " packets would be dropped, " +
+          "but logs only report " + droppedPackets + " dropped packets");
```
Comment (islemaster) [inline apps/test/netsim/NetSimRouterNode.js]: Style guide nit: use 4-space indentation for line continuation.

File: apps/test/netsim/dataConverters.js [hunk 10]
```diff
@@ -632,6 +632,123 @@ describe("dataConverters", function () {
...
+  describe("binaryToBase64", function () {
+    var binaryToBase64 = dataConverters.binaryToBase64;
+    var base64ToBinary = dataConverters.base64ToBinary;
+
+    /**
+    * Assert if the given base64 results are not equal. Equality is
+    * defined as sharing BOTH their string and their len properties
+    * @param {Object} left
+    * @param {Object} right
+    */
```
Comment (islemaster) [inline apps/test/netsim/dataConverters.js]: Nit: JSdoc stars don't line up :trollface:

**Comment by islemaster** [comment]
Beautifully done.  Thanks for being so diligent about renaming things appropriately, and taking on the pain of my many early assumptions about the messageRow type.

I see you're already underway responding to comments.  I'll scan one more time when you're done.  It may be prudent to wait until the morning to merge.

_Pulled and passed tests locally_

Nice work!

File: apps/src/netsim/NetSimMessage.js (hunk 3 above)
Comment (islemaster) [inline apps/src/netsim/NetSimMessage.js]: We don't have a style guide rule about it, feel free to keep if you prefer them.

**Comment by Hamms** [comment]
Hooray, thank you both! I believe the last two commits took care of the remaining comments, and I'm excited to merge as soon as you complete your final check.

**Comment by islemaster** [comment]
:+1: LGTM.  Merge it!

**Pull request merged**